Sources/PenguinStructures/ArrayStorage.swift

```diff
@@ -0,0 +1,184 @@
+  private var access: Accessor { .init(unsafeBufferObject: self) }
+
+  /// The number of elements stored in `self`.
+  public var count: Int {
+    _read { yield access.withUnsafeMutablePointerToHeader { $0.pointee.count } }
+    _modify {
+      defer { _fixLifetime(self) }
+      yield &access.withUnsafeMutablePointerToHeader { $0 }.pointee.count
+    }
+  }
+
+  /// The maximum number of elements that can be stored in `self`.
+  public var capacity: Int {
+    _read {
+      yield access.withUnsafeMutablePointerToHeader {
+        $0.pointee.capacity }
+    }
+    _modify {
+      defer { _fixLifetime(self) }
+      yield &access.withUnsafeMutablePointerToHeader { $0 }.pointee.capacity
+    }
+  }
+  
+  /// Appends `x` if possible, returning the index of the appended element or
+  /// `nil` if there was insufficient capacity remaining.
+  public func append(_ x: Element) -> Int? {
+    let r = count
+    if r == capacity { return nil }
+    access.withUnsafeMutablePointers { h, e in
+      (e + r).initialize(to: x)
+      h[0].count = r + 1
+    }
+    return r
+  }
+
+  /// Invokes `body` with the memory occupied by stored elements.
+  public func withUnsafeMutableBufferPointer<R>(
+    _ body: (inout UnsafeMutableBufferPointer<Element>) -> R
+  ) -> R {
+    access.withUnsafeMutablePointers { h, e in
+      var b = UnsafeMutableBufferPointer(start: e, count: h[0].count)
+      return body(&b)
+    }
+  }
+
+  /// Returns an empty instance with `capacity` at least `minimumCapacity`.
+  public static func create(minimumCapacity: Int) -> Self {
+    unsafeDowncast(
+      Accessor(bufferClass: Self.self, minimumCapacity: minimumCapacity) {
+        buffer, getCapacity in 
+        ArrayHeader(count: 0, capacity: getCapacity(buffer))
+      }.buffer,
+      to: Self.self)
+  }
+
+  private init() { fatalError("Please call create()") }
+}
+
+/// Implementation of `AnyArrayStorageImplementation` requirements
+extension ArrayStorageImplementation {
+  /// Appends the instance of the concrete element type whose address is `p`,
+  /// returning the index of the appended element, or `nil` if there was
+  /// insufficient capacity remaining
+  public func appendValue_(at p: UnsafeRawPointer) -> Int? {
+    append(p.assumingMemoryBound(to: Element.self)[0])
+  }
+
+  /// Invokes `body` with the memory occupied by initialized elements.
+  public func withUnsafeMutableRawBufferPointer_<R>(
+    _ body: (inout UnsafeMutableRawBufferPointer)->R
+  ) -> R {
+    withUnsafeMutableBufferPointer {
+      var b = UnsafeMutableRawBufferPointer($0)
+      return body(&b)
+    }
+  }
+
+  /// Deinitialize stored data. Models should call this from their `deinit`.
+  public func deinitialize() {
+    access.withUnsafeMutablePointers { h, e in
+      e.deinitialize(count: h[0].count)
+      h.deinitialize(count: 1)
+    }
+  }
+}
+
+/// Type-erasable storage for contiguous `Factoid` `Element` instances.
```

Comment:
Is this Factor-graph specific?`Factoid` seems to only exist in the tests?

Reply:
It's not Factor-graph specific, but it's what's needed to support the Factor-graph use case.  `Factoid` is a stand-in for `Factor` to show that we could adapt this code to a similar situation.

Reply:
Since `Factoid` is not defined or used anywhere except for the tests, could you consider just removing it in this comment? Otherwise, LGTM!

Reply:
Oh, definitely, this comment shouldn't be here.  Good catch!

Reply:
41163ab